assertsprocessor/config_refresh.go

```diff
@@ -0,0 +1,83 @@
+package assertsprocessor
+
+import (
+	"encoding/json"
+	"github.com/tilinna/clock"
+	"go.uber.org/zap"
+	"net/http"
+)
+
+type configListener interface {
+	isUpdated(currConfig *Config, latestConfig *Config) bool
+	onUpdate(config *Config) error
+}
+
+type configRefresh struct {
+	config           *Config
+	logger           *zap.Logger
+	configSyncTicker *clock.Ticker
+	stop             chan bool
+	assertsClient    *assertsClient
+	spanMatcher      *spanMatcher
+	configListeners  []configListener
+}
+
+func (cr *configRefresh) stopUpdates() {
+	go func() { cr.stop <- true }()
+}
+
+func (cr *configRefresh) startUpdates() {
+	endPoint := (*(*cr.config).AssertsServer)["endpoint"]
+	if endPoint != "" {
+		go func() {
+			for {
+				select {
+				case <-cr.stop:
+					cr.logger.Info("Stopping collector config updates")
+					return
+				case <-cr.configSyncTicker.C:
+					cr.logger.Info("Fetching collector config")
+					cr.fetchAndUpdateConfig()
+				}
+			}
+		}()
+	}
+}
+
+func (cr *configRefresh) fetchAndUpdateConfig() {
+	latestConfig, err := cr.fetchConfig()
+	if err == nil {
+		cr.updateConfig(latestConfig)
+	}
+}
+
+func (cr *configRefresh) fetchConfig() (*Config, error) {
+	var config *Config
+	body, err := cr.assertsClient.invoke(http.MethodGet, configApi, nil)
+	if err == nil {
+		err = json.Unmarshal(body, config)
+		if err == nil {
+			cr.logConfig(config)
+		} else {
+			cr.logger.Error("Error unmarshalling config", zap.Error(err))
+		}
+	}
+
+	return config, err
+}
+
+func (cr *configRefresh) updateConfig(latestConfig *Config) {
+	err := error(nil)
+	for _, listener := range cr.configListeners {
+		if listener.isUpdated(cr.config, latestConfig) {
+			err = listener.onUpdate(latestConfig)
+		}
+	}
+	if err == nil {
+		cr.config = latestConfig
```

Comment:
The config updates are accepted only when all the listeners  consume the config updates without error. But this could leave different components with different versions of config. This is fine for now, but better to call it out in a comment.

Reply:
I spent a bit of time thinking about this. This is like an XA commit problem :-) I'll go with your suggestion for now